draft-duke-quic-load-balancers.md

```diff
@@ -505,6 +518,208 @@ reserved for config rotation {{config-rotation}}.
+of mode MAY be made on a per-packet basis, through a stochastic process or based
+on client address.
+
+A retry service MUST forward all packets for a QUIC version it does not
+understand. Note that if servers support versions the retry service does not,
+this may unacceptably increase loads on the servers. However, dropping these
+packets would introduce chokepoints to block deployment of new QUIC versions.
+
+## No-Shared-State Retry Service
+
+The no-shared-state retry service requires no coordination, except that the
+server must be configured to accept this service. The scheme uses the first bit
+of the token to distinguish between tokens from Retry packets (codepoint '0')
+and tokens from NEW_TOKEN frames (codepoint '1').
+
+### Service Requirements {#nss-service-requirements}
+
+A no-shared-state retry service MUST be present on all paths from the server to
+potential clients. These paths MUST fail to pass QUIC traffic should the service
+fail for any reason. That is, if the service is not operational, the server MUST
+NOT be exposed to client traffic. Otherwise, servers that have already disabled
+their Retry capability would be vulnerable to attack.
+
+The path between service and server MUST be free of any potential attackers. Note
+that this and other requirements above severely restrict the operational
+conditions in which a no-shared-state retry service can safely operate.
+
+Retry tokens generated by the service MUST have the format below.
+
+~~~~~
+0                   1                   2                   3
+0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|0| ODCIL (7) |  Original Destination Connection ID (0..160)    |      
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|           Original Destination Connection ID (...)            |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+...
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                     Opaque Data (variable)                    |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+{: #nss-retry-service-token-format title="Format of non-shared-state retry service tokens"}
+
+The first bit of retry tokens generated by the service must be zero. The token
+has the following additional fields:
+
+ODCIL: The length of the original destination connection ID from the triggering
+Initial packet. This is in cleartext to be readable for the server, but
+authenticated later in the token.
+
+Original Destination Connection ID: This also in cleartext and authenticated
+later.
+
+Opaque Data: This data MUST contain encrypted information that allows the retry
+service to validate the client's IP address, in accordance with the QUIC
+specification. It MUST also encode a secure hash of the original destination
+connection ID field to verify that this field has not been edited.
```

Comment:
Should we state something about ensuring a small time windows in which the token is valid?

Reply:
I think "in accordance with the QUIC specification" covers all of those issues.

Reply:
Does the QUIC spec state anything about time? I still think it would be worthwhile text to have here.

Reply:
If there's a problem, it's with the QUIC spec. I don't think there are any special requirements for QUIC-LB besides what's listed here.